sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,287 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/stdasm.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * @file
+ * @brief Control and Status Register (CSR) interface.
+ *
+ * A set of macros that provide both read and modify operations on CSRs.
+ * Compiling translation units that include this header file with `-DMOCK_CSR`
+ * will result in the CSR operations being replaced with a mocked
+ * implementation.
+ */
+
+/**
+ * Define the implementation macros.
+ *
+ * The implementation used depends on whether the CSR library is providing a
+ * real or a mocked interface.
+ */
+#ifdef MOCK_CSR
+
+uint32_t mock_csr_read(uint32_t addr);
+
+#define CSR_READ_IMPL(csr, dst)  \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
```

Comment:
Perhaps worth adding a comment noting that this is to ensure that `csr` is a constant?

Reply:
Added a commented macro.